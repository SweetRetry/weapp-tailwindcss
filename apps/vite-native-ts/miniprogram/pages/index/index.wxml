<view class="p-4">
  <IceButton class="bg-[#d72929]" my-class="bg-[orange]"></IceButton>
</view>