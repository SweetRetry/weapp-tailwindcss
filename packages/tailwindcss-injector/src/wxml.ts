import type { Buffer } from 'node:buffer'
import path from 'node:path'
import fs from 'fs-extra'
import { Parser } from 'htmlparser2'
import MagicString from 'magic-string'

const srcImportTagsMap: Record<string, string[]> = {
  // audio: ['src', 'poster'],
  // video: ['src', 'poster'],
  // image: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxs/01wxs-module.html
  // wxs: ['src'],
  // https://developers.weixin.qq.com/miniprogram/dev/reference/wxml/import.html
  import: ['src'],
  include: ['src'],
}

export interface WxmlDep {
  tagName: string
  start: number
  end: number
  quote: string | null | undefined
  name: string
  value: string
  attrs: Record<string, string>
}

// https://github.com/fb55/htmlparser2/issues/1541
// https://developers.weixin.qq.com/miniprogram/dev/framework/view/wxml/event.html
export function processWxml(wxml: string | Buffer) {
  const ms = new MagicString(wxml.toString())
  const deps: WxmlDep[] = []
  let currentTagName = ''
  let importAttrs: undefined | string[]
  let attrs: Record<string, string> = {}

  const parser = new Parser(
    {
      onopentagname(name) {
        currentTagName = name
        importAttrs = srcImportTagsMap[currentTagName]
      },
      onattribute(name, value, quote) {
        attrs[name] = value
        if (importAttrs) {
          for (const attrName of importAttrs) {
            if (attrName === name) {
              deps.push({
                name,
                value,
                quote,
                tagName: currentTagName,
                start: parser.startIndex,
                end: parser.endIndex,
                attrs,
              })
            }
          }
        }
      },
      onclosetag() {
        currentTagName = ''
        attrs = {}
        importAttrs = undefined
      },
    },
    {
      lowerCaseTags: false,
      xmlMode: true,
    },
  )
  parser.write(
    ms.original,
  )
  parser.end()
  return deps
}

export async function getDepFiles(file: string) {
  const set = new Set<string>()

  async function getDep(filepath: string) {
    if (await fs.exists(filepath)) {
      const dirname = path.dirname(filepath)
      set.add(filepath)
      const wxml = await fs.readFile(filepath, 'utf8')
      const deps = await processWxml(wxml)
      for (const { value } of deps) {
        const p = path.resolve(dirname, value)
        if (!set.has(p)) {
          await getDep(p)
        }
      }
    }
  }

  await getDep(file)

  return set
}
