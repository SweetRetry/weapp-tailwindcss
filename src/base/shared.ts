import groupBy from 'lodash.groupby'
import type { UserDefinedOptions } from '@/types'

export function getGroupedEntries<T> (entries: [string, T][], options: Required<UserDefinedOptions>) {
  const { cssMatcher, htmlMatcher, jsMatcher } = options
  const groupedEntries = groupBy(entries, ([file]) => {
    if (cssMatcher(file)) {
      return 'css'
    } else if (htmlMatcher(file)) {
      return 'html'
    } else if (jsMatcher(file)) {
      return 'js'
    } else {
      return 'other'
    }
  })
  return groupedEntries
}
