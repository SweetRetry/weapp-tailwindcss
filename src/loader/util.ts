import fs from 'fs'
import path from 'path'

export function mkfileSync (filename: string, content: string) {
  try {
    const dirname = path.dirname(filename)
    if (!fs.existsSync(dirname)) {
      fs.mkdirSync(dirname, {
        recursive: true
      })
    }
    fs.writeFileSync(filename, content, 'utf-8')
  } catch (error) {
    console.error(error)
  }
}
