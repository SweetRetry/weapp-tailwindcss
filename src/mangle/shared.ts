export const mangleClassPrefix = 'MANGLE__'

export const mangleClassSuffix = '__MANGLE'

export const mangleClassRegex = /MANGLE__[a-zA-Z0-9_-]+__MANGLE/g

export function markForMangled (str: string) {
  if (typeof str === 'string' && str) {
    return `${mangleClassPrefix}${str.trim()}${mangleClassSuffix}`
  }
  return str
}
