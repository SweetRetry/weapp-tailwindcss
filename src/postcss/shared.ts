import { escape } from '@/escape'
import { SimpleMappingChars2String } from '@/dic'
// css 中，要多加一个 '\' 来转义
// for raw css selector
// export function cssSelectorReplacer(selector: string, escapeEntries = MappingChars2StringEntries) {
//   return escape(selector, true, escapeEntries).replace(/\\2c /g, dic[','])
// }

export function internalCssSelectorReplacer(selectors: string, map: Record<string, string> = SimpleMappingChars2String) {
  return escape(selectors, {
    map
  })
}
