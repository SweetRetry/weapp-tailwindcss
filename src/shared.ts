import path from 'path'

export { styleHandler } from './postcss'
export { templeteHandler } from './wxml'
export { jsxHandler } from './jsx'
export const pluginName = 'weapp-tailwindcss-webpack-plugin'

export function getFileName (file: string) {
  return path.basename(file, path.extname(file))
}
