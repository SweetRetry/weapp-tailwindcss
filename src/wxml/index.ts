import { parse, render } from 'wxml-ast'
import { parseExpression } from '@babel/parser'
import traverse from '@babel/traverse'
import generate from '@babel/generator'
import { replaceWxml } from './shared'

export { replaceWxml }

// #region internal
const variableRegExp = /{{([^{}]*)}}/g

function variableMatch(original: string) {
  return variableRegExp.exec(original)
}
// #endregion

export function generateCode(match: string) {
  const ast = parseExpression(match)

  traverse(ast, {
    StringLiteral(path) {
      path.node.value = replaceWxml(path.node.value)
    },
    noScope: true
  })

  const { code } = generate(ast, {
    compact: true,
    minified: true
  })
  return code
}

export interface RawSource {
  start: number
  end: number
  raw: string
  // '' 直接 remove {{}}
  source?: string
}

export function templeteReplacer(original: string) {
  let match = variableMatch(original)
  const sources: RawSource[] = []

  while (match !== null) {
    // 过滤空字符串
    // if (match[1].trim().length) {
    sources.push({
      start: match.index,
      end: variableRegExp.lastIndex,
      raw: match[1]
    })

    match = variableMatch(original)
  }

  if (sources.length) {
    const resultArray: string[] = []
    let p = 0
    for (let i = 0; i < sources.length; i++) {
      const m = sources[i]
      // 匹配前值
      resultArray.push(replaceWxml(original.slice(p, m.start), true))
      p = m.start
      // 匹配后值
      if (m.raw.trim().length) {
        const code = generateCode(m.raw)
        m.source = `{{${code}}}`
      } else {
        m.source = ''
      }

      resultArray.push(m.source)
      p = m.end
      // 匹配最终尾部值
      if (i === sources.length - 1) {
        resultArray.push(replaceWxml(original.slice(m.end), true))
      }
    }

    return resultArray.filter((x) => x).join('')
  } else {
    return replaceWxml(original)
  }
}

export async function templeteHandler(rawSource: string) {
  const parsed = await parse(rawSource, {
    elementCB: (node) => {
      if (node.attribs.class) {
        node.attribs.class = templeteReplacer(node.attribs.class)
      }
    }
  })
  return render(parsed)
  // wxml.traverse(parsed, (node, parent) => {
  //   if (node.type === wxml.NODE_TYPES.ELEMENT) {
  //     // @ts-ignore
  //     if (node.attributes.class) {
  //       // @ts-ignore
  //       node.attributes.class = templeteReplacer(node.attributes.class)
  //     }
  //   }
  // })
  // return wxml.serialize(parsed)
}
