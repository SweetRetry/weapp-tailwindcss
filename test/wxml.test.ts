import { templeteReplacer, replaceWxml, templeteHandler } from '@/wxml/index'
import { classStringReplace } from '@/shared'
import { wxmlCasePath, createGetCase, createPutCase } from './util'
const getCase = createGetCase(wxmlCasePath)
// @ts-ignore
// eslint-disable-next-line no-unused-vars
const putCase = createPutCase(wxmlCasePath)
describe('wxml', () => {
  it('isStringLiteral', () => {
    const testCase = "{{['som-node__label','data-v-59229c4a','som-org__text-'+(node.align||''),node.active||collapsed?'som-node__label-active':'',d]}}"

    const result = templeteReplacer(testCase)

    expect(result).toBe('{{["som-node__label","data-v-59229c4a","som-org__text-"+(node.align||""),node.active||collapsed?"som-node__label-active":"",d]}}')
    expect(result).toMatchSnapshot()
  })

  it('isConditionalExpression', () => {
    const testCase = "{{['flex','flex-col','items-center',flag===1?'bg-red-900':'bg-[#fafa00]']}}"
    const result = templeteReplacer(testCase)
    expect(result).toBe('{{["flex","flex-col","items-center",flag===1?"bg-red-900":"bg-_l__h_fafa00_r_"]}}')
    expect(result).toMatchSnapshot()
  })

  it('nest ', () => {
    const testCase =
      "{{[flag?'bg-red-900':'bg-[#fafa00]',classObject,[(flag===true)?'bg-[#fafa00]':'',(true)?'text-sm':''],flag?flag===false?'bg-red-900':'bg-[#000]':'bg-[#fafa00]']}}"

    const result = templeteReplacer(testCase)
    expect(result).toBe(
      '{{[flag?"bg-red-900":"bg-_l__h_fafa00_r_",classObject,[flag===true?"bg-_l__h_fafa00_r_":"",true?"text-sm":""],flag?flag===false?"bg-red-900":"bg-_l__h_000_r_":"bg-_l__h_fafa00_r_"]}}'
    )
    expect(result).toMatchSnapshot()
  })

  it('sm:text-3xl dark:text-sky-400', () => {
    const testCase = 'sm:text-3xl dark:text-slate-200 bg-[#ffffaa]'
    const result = templeteReplacer(testCase)
    expect(result).toBe('sm_c_text-3xl dark_c_text-slate-200 bg-_l__h_ffffaa_r_')
    expect(result).toMatchSnapshot()
  })
  // it('should ', () => {
  //   const case3 =
  //     '<view class="_div data-v-59d503f6"><view class="container data-v-59d503f6">container</view><label class="decoration-clone bg-gradient-to-b from-yellow-400 to-red-500 text-transparent _span data-v-59d503f6">Hello<view class="_br data-v-59d503f6"></view>World</label><view class="box-border h-32 w-32 p-4 border-4 border-blue-400 bg-blue-200 rounded-md _div data-v-59d503f6"><view class="h-full w-full bg-blue-400 bg-stripes bg-stripes-white _div data-v-59d503f6"></view></view><view class="box-content h-32 w-32 p-4 border-4 border-green-400 bg-green-200 rounded-md _div data-v-59d503f6"><view class="h-full w-full bg-green-400 bg-stripes bg-stripes-white _div data-v-59d503f6"></view></view><block wx:for="{{displayArray}}" wx:for-item="d" wx:for-index="__i0__" wx:key="*this"><view class="{{_l_\'item\',\'data-v-59d503f6\',d]}}">{{d}}</view></block></view>'
  // })

  it('static node = self', () => {
    const testCase =
      '<view class="p-[20px] -mt-2 mb-[-20px]">p-[20px] -mt-2 mb-[-20px] margin的jit 不能这么写 -m-[20px]</view><view class="space-y-[1.6rem]"><view class="w-[300rpx] text-black text-opacity-[0.19]">w-[300rpx] text-black text-opacity-[0.19]</view><view class="min-w-[300rpx] max-h-[100px] text-[20px] leading-[0.9]">min-w-[300rpx] max-h-[100px] text-[20px] leading-[0.9]</view><view class="max-w-[300rpx] min-h-[100px] text-[#dddddd]">max-w-[300rpx] min-h-[100px] text-[#dddddd]</view><view class="flex items-center justify-center h-[100px] w-[100px] rounded-[40px] bg-[#123456] bg-opacity-[0.54] text-[#ffffff]">Hello</view><view class="border-[10px] border-[#098765] border-solid border-opacity-[0.44]">border-[10px] border-[#098765] border-solid border-opacity-[0.44]</view><view class="grid grid-cols-3 divide-x-[10px] divide-[#010101] divide-solid"><view>1</view><view>2</view><view>3</view></view></view><view class="test">test</view>'
    const result = classStringReplace(testCase, (x) => x)
    expect(result).toBe(testCase)
    expect(result).toMatchSnapshot()
  })

  it('static node ', () => {
    const testCase =
      '<view class="p-[20px] -mt-2 mb-[-20px]">p-[20px] -mt-2 mb-[-20px] margin的jit 不能这么写 -m-[20px]</view><view class="space-y-[1.6rem]"><view class="w-[300rpx] text-black text-opacity-[0.19]">w-[300rpx] text-black text-opacity-[0.19]</view><view class="min-w-[300rpx] max-h-[100px] text-[20px] leading-[0.9]">min-w-[300rpx] max-h-[100px] text-[20px] leading-[0.9]</view><view class="max-w-[300rpx] min-h-[100px] text-[#dddddd]">max-w-[300rpx] min-h-[100px] text-[#dddddd]</view><view class="flex items-center justify-center h-[100px] w-[100px] rounded-[40px] bg-[#123456] bg-opacity-[0.54] text-[#ffffff]">Hello</view><view class="border-[10px] border-[#098765] border-solid border-opacity-[0.44]">border-[10px] border-[#098765] border-solid border-opacity-[0.44]</view><view class="grid grid-cols-3 divide-x-[10px] divide-[#010101] divide-solid"><view>1</view><view>2</view><view>3</view></view></view><view class="test">test</view>'
    const result = classStringReplace(testCase, (x) => {
      return replaceWxml(x)
    })
    expect(result).toBe(
      '<view class="p-_l_20px_r_ -mt-2 mb-_l_-20px_r_">p-[20px] -mt-2 mb-[-20px] margin的jit 不能这么写 -m-[20px]</view><view class="space-y-_l_1-dot-6rem_r_"><view class="w-_l_300rpx_r_ text-black text-opacity-_l_0-dot-19_r_">w-[300rpx] text-black text-opacity-[0.19]</view><view class="min-w-_l_300rpx_r_ max-h-_l_100px_r_ text-_l_20px_r_ leading-_l_0-dot-9_r_">min-w-[300rpx] max-h-[100px] text-[20px] leading-[0.9]</view><view class="max-w-_l_300rpx_r_ min-h-_l_100px_r_ text-_l__h_dddddd_r_">max-w-[300rpx] min-h-[100px] text-[#dddddd]</view><view class="flex items-center justify-center h-_l_100px_r_ w-_l_100px_r_ rounded-_l_40px_r_ bg-_l__h_123456_r_ bg-opacity-_l_0-dot-54_r_ text-_l__h_ffffff_r_">Hello</view><view class="border-_l_10px_r_ border-_l__h_098765_r_ border-solid border-opacity-_l_0-dot-44_r_">border-[10px] border-[#098765] border-solid border-opacity-[0.44]</view><view class="grid grid-cols-3 divide-x-_l_10px_r_ divide-_l__h_010101_r_ divide-solid"><view>1</view><view>2</view><view>3</view></view></view><view class="test">test</view>'
    )
    expect(result).toMatchSnapshot()
  })

  it('percentage unit', () => {
    const testCase = '<view class="h-[200%]" />'
    const result = classStringReplace(testCase, (x) => {
      return replaceWxml(x)
    })
    expect(result).toBe('<view class="h-_l_200_pct__r_" />')
  })

  it('\\r\\n replace test', async () => {
    const testCase = `
    bg-white
    rounded-full
    w-10
    h-10
    flex
    justify-center
    items-center
    pointer-events-auto
  `
    const result = templeteReplacer(testCase)
    expect(result).toBe('bg-white rounded-full w-10 h-10 flex justify-center items-center pointer-events-auto')
  })

  it('\\r\\n replace test with var', async () => {
    const testCase = `{{[
      'flex',
      'items-center',
      'justify-center',
      'h-_l_100px_r_',
      'w-_l_100px_r_',
      'rounded-_l_40px_r_',
      'bg-_l__h_123456_r_',
      'bg-opacity-_l_0-dot-54_r_',
      'text-_l__h_ffffff_r_',
      'data-v-1badc801',
      'text-_l__h_123456_r_',
      b]}}`
    const result = templeteReplacer(testCase)
    expect(result).toBe(
      '{{["flex","items-center","justify-center","h-_l_100px_r_","w-_l_100px_r_","rounded-_l_40px_r_","bg-_l__h_123456_r_","bg-opacity-_l_0-dot-54_r_","text-_l__h_ffffff_r_","data-v-1badc801","text-_l__h_123456_r_",b]}}'
    )
  })
})
// bg-[rgb(2,132,199)]
