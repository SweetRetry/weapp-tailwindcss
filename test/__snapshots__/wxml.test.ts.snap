// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wxml isConditionalExpression 1`] = `"{{['flex','flex-col','items-center',flag===1?'bg-red-900':'bg-_l__h_fafa00_r_']}}"`;

exports[`wxml isStringLiteral 1`] = `"{{['som-node__label','data-v-59229c4a','som-org__text-'+(node.align||''),node.active||collapsed?'som-node__label-active':'',d]}}"`;

exports[`wxml nest  1`] = `"{{[flag?'bg-red-900':'bg-_l__h_fafa00_r_',classObject,[flag===true?'bg-_l__h_fafa00_r_':'',true?'text-sm':''],flag?flag===false?'bg-red-900':'bg-_l__h_000_r_':'bg-_l__h_fafa00_r_']}}"`;

exports[`wxml sm:text-3xl dark:text-sky-400 1`] = `"sm_c_text-3xl dark_c_text-slate-200 bg-_l__h_ffffaa_r_"`;

exports[`wxml static node  1`] = `"<view class=\\"p-_l_20px_r_ -mt-2 mb-_l_-20px_r_\\">p-[20px] -mt-2 mb-[-20px] margin的jit 不能这么写 -m-[20px]</view><view class=\\"space-y-_l_1-dot-6rem_r_\\"><view class=\\"w-_l_300rpx_r_ text-black text-opacity-_l_0-dot-19_r_\\">w-[300rpx] text-black text-opacity-[0.19]</view><view class=\\"min-w-_l_300rpx_r_ max-h-_l_100px_r_ text-_l_20px_r_ leading-_l_0-dot-9_r_\\">min-w-[300rpx] max-h-[100px] text-[20px] leading-[0.9]</view><view class=\\"max-w-_l_300rpx_r_ min-h-_l_100px_r_ text-_l__h_dddddd_r_\\">max-w-[300rpx] min-h-[100px] text-[#dddddd]</view><view class=\\"flex items-center justify-center h-_l_100px_r_ w-_l_100px_r_ rounded-_l_40px_r_ bg-_l__h_123456_r_ bg-opacity-_l_0-dot-54_r_ text-_l__h_ffffff_r_\\">Hello</view><view class=\\"border-_l_10px_r_ border-_l__h_098765_r_ border-solid border-opacity-_l_0-dot-44_r_\\">border-[10px] border-[#098765] border-solid border-opacity-[0.44]</view><view class=\\"grid grid-cols-3 divide-x-_l_10px_r_ divide-_l__h_010101_r_ divide-solid\\"><view>1</view><view>2</view><view>3</view></view></view><view class=\\"test\\">test</view>"`;

exports[`wxml static node = self 1`] = `"<view class=\\"p-[20px] -mt-2 mb-[-20px]\\">p-[20px] -mt-2 mb-[-20px] margin的jit 不能这么写 -m-[20px]</view><view class=\\"space-y-[1.6rem]\\"><view class=\\"w-[300rpx] text-black text-opacity-[0.19]\\">w-[300rpx] text-black text-opacity-[0.19]</view><view class=\\"min-w-[300rpx] max-h-[100px] text-[20px] leading-[0.9]\\">min-w-[300rpx] max-h-[100px] text-[20px] leading-[0.9]</view><view class=\\"max-w-[300rpx] min-h-[100px] text-[#dddddd]\\">max-w-[300rpx] min-h-[100px] text-[#dddddd]</view><view class=\\"flex items-center justify-center h-[100px] w-[100px] rounded-[40px] bg-[#123456] bg-opacity-[0.54] text-[#ffffff]\\">Hello</view><view class=\\"border-[10px] border-[#098765] border-solid border-opacity-[0.44]\\">border-[10px] border-[#098765] border-solid border-opacity-[0.44]</view><view class=\\"grid grid-cols-3 divide-x-[10px] divide-[#010101] divide-solid\\"><view>1</view><view>2</view><view>3</view></view></view><view class=\\"test\\">test</view>"`;
